safekeeper/src/bin/safekeeper.rs

```diff
@@ -220,71 +230,77 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
 
-    threads.push(safekeeper_thread);
+    let conf_ = conf.clone();
+    let http_handle = HTTP_RUNTIME
+        .spawn(http::task_main(conf_, http_listener))
+        .map(|res| ("HTTP service main".to_owned(), res));
+    tasks_handles.push(Box::pin(http_handle));
 
     let conf_ = conf.clone();
-    threads.push(
-        thread::Builder::new()
-            .name("broker thread".into())
-            .spawn(|| {
-                broker::thread_main(conf_);
-            })?,
-    );
+    let broker_task_handle = BROKER_RUNTIME
+        .spawn(broker::task_main(conf_).instrument(info_span!("broker")))
+        .map(|res| ("broker main".to_owned(), res));
+    tasks_handles.push(Box::pin(broker_task_handle));
 
     let conf_ = conf.clone();
-    threads.push(
-        thread::Builder::new()
-            .name("WAL removal thread".into())
-            .spawn(|| {
-                remove_wal::thread_main(conf_);
-            })?,
-    );
+    let wal_remover_handle = WAL_REMOVER_RUNTIME
+        .spawn(remove_wal::task_main(conf_))
+        .map(|res| ("WAL remover".to_owned(), res));
+    tasks_handles.push(Box::pin(wal_remover_handle));
 
-    threads.push(
-        thread::Builder::new()
-            .name("WAL backup launcher thread".into())
-            .spawn(move || {
-                wal_backup::wal_backup_launcher_thread_main(conf, wal_backup_launcher_rx);
-            })?,
-    );
+    let conf_ = conf.clone();
+    let wal_backup_handle = WAL_BACKUP_RUNTIME
+        .spawn(wal_backup::wal_backup_launcher_task_main(
+            conf_,
+            wal_backup_launcher_rx,
+        ))
+        .map(|res| ("WAL backup launcher".to_owned(), res));
+    tasks_handles.push(Box::pin(wal_backup_handle));
+
+    let metrics_shifter_handle = METRICS_SHIFTER_RUNTIME
+        .spawn(safekeeper::metrics::metrics_shifter())
+        .map(|res| ("metrics shifter".to_owned(), res));
+    tasks_handles.push(Box::pin(metrics_shifter_handle));
 
     set_build_info_metric(GIT_VERSION);
-    // TODO: put more thoughts into handling of failed threads
-    // We should catch & die if they are in trouble.
-
-    // On any shutdown signal, log receival and exit. Additionally, handling
-    // SIGQUIT prevents coredump.
-    ShutdownSignals::handle(|signal| {
-        info!("received {}, terminating", signal.name());
-        std::process::exit(0);
-    })
+
+    // TODO: update tokio-stream, convert to real async Stream with
+    // SignalStream, map it to obtain missing signal name, combine streams into
+    // single stream we can easily sit on.
+    let mut sigquit_stream = signal(SignalKind::quit())?;
+    let mut sigint_stream = signal(SignalKind::interrupt())?;
+    let mut sigterm_stream = signal(SignalKind::terminate())?;
+
+    tokio::select! {
+        Some((task_name, res)) = tasks_handles.next()=> {
+            error!("{} task failed: {:?}, exiting", task_name, res);
+            std::process::exit(1);
```

Comment:
This will skip all of the drop handlers, which might be surprising while trying to debug.

Would recommend either returning a exit code from this fn or an `Error` to be logged by `main`.

With returning, it's good to keep in mind that by using `FuturesUnordered` or anything else than a `JoinSet`, you will only see a random first error here but the spawned tasks will continue to run until in the background regardless of exiting this method.

Still, I think it would be most sensible to allow dropping or dtors to run. Because if someone adds a dropguard anywhere, and it doesn't run because of some other task failed and bubbled up to these exits, you will have zero logging on that and no way to confirm it running or not.

Reply:
Again I initially missed that `FuturesUnordered` is used to turn every `Ok(())` into an `Err`, so try_join! will not work without additional massaging.

Reply:
Cannot see a way to do this with joinset, but here's the try_join idea: https://github.com/neondatabase/neon/pull/4124/commits/4d96a27e3ec5f4e933c1482e0dc65fc7ed24d31c

Reply:
Safekeepers are not the complicated. There is no graceful shutdown of any kinds and no drops which are expected to be run on shutdown, and I prefer to keep things simple unless otherwise is required, so this seems to be a bit premature.

Reply:
This is different from graceful shutdown, it's about "clean" as expected in rust applications shutdown, so that destructors are run (when not doing abort). I don't know if it was in place before, but now would be a good time to add it. It might expose some deadlocks at worst but systemd will send a SIGKILL eventually.

Reply:
But do you have any idea of concrete examples where this would be needed?

This is one of the areas where I personally much more bothered by other things, but don't mind someone picking it up.

Reply:
btw, I don't see what try_join changes here; stopping polling handles doesn't mean tasks are stopped. Doing return instead of std::process::exit is possible with older version as well. Not sure what are rust guarantees when main thread finishes; will it call drops in other threads? Probably yes.

Reply:
Aah yes, actually dropping with multiple runtimes... I don't know what happens. The runtime running `async fn main` will get dropped at least.

But, ok. Don't want to hold this.